doc/man/man3/thread_pair.3
.TH "thread_pair" 3 "Wed May 10 2023" "Version 01.00" "Dispatcher TCP/IP" \" -*- nroff -*-
.ad l
.nh
.SH NAME
thread_pair
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <thread_pair\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBthread_pair\fP ()=delete"
.br
.ti -1c
.RI "\fBthread_pair\fP (\fBMessageQueue\fP write_queue_id, \fBMessageQueue\fP common_queue, int idx, std::shared_ptr< \fBkeep_running_flags\fP > shpt_keep_running, std::shared_ptr< \fBconnections\fP > shpt_conn, std::shared_ptr< \fBSemaphore\fP > shpt_sem)"
.br
.ti -1c
.RI "\fBthread_pair\fP (const \fBthread_pair\fP &)"
.br
.ti -1c
.RI "\fBthread_pair\fP (\fBthread_pair\fP &&)=delete"
.br
.ti -1c
.RI "\fBthread_pair\fP & \fBoperator=\fP (const \fBthread_pair\fP &)=default"
.br
.ti -1c
.RI "\fBthread_pair\fP & \fBoperator=\fP (\fBthread_pair\fP &&)=delete"
.br
.ti -1c
.RI "void \fBreader_thread\fP (int idx_thp)"
.br
.ti -1c
.RI "void \fBwriter_thread\fP (int idx_thp)"
.br
.ti -1c
.RI "int \fBadd_sockdata\fP (\fBsocket_data_t\fP sdt)"
.br
.ti -1c
.RI "int \fBremove_sockdata\fP (const \fBsocket_data_t\fP &sdt)"
.br
.ti -1c
.RI "int \fBget_sockdata_list\fP (std::list< \fBsocket_data_t\fP > &lsdt)"
.br
.ti -1c
.RI "int \fBget_size_of_sock_list\fP () const"
.br
.ti -1c
.RI "int \fBget_read_pipe\fP () const"
.br
.ti -1c
.RI "int \fBget_write_pipe\fP () const"
.br
.ti -1c
.RI "int \fBget_id\fP () const"
.br
.ti -1c
.RI "int \fBget_idx\fP () const"
.br
.ti -1c
.RI "\fBMessageQueue\fP & \fBget_write_queue\fP ()"
.br
.ti -1c
.RI "void \fBPrepare_Msg_Json_To_Send\fP (\fBprotomsg::st_protomsg\fP &v_protomsg, std::string msg, Json::Value &json_msg)"
.br
.ti -1c
.RI "void \fBAttending_Read_Socket\fP (\fBsocket_data_t\fP &sdt)"
.br
.ti -1c
.RI "int \fBGetting_Json_Msg_Received\fP (std::string &msgin, \fBprotomsg::st_protomsg\fP &v_protomsg, std::string &msgout)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "std::thread \fBth_r\fP"
.br
.ti -1c
.RI "std::thread \fBth_w\fP"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "int \fB_idx_thp\fP {\-1}"
.br
.ti -1c
.RI "int \fB_pipe\fP [2] {\-1,\-1}"
.br
.ti -1c
.RI "\fBMessageQueue\fP \fB_write_queue\fP"
.br
.ti -1c
.RI "\fBMessageQueue\fP \fB_common_queue\fP"
.br
.ti -1c
.RI "std::mutex \fB_accept_mutex\fP"
.br
.ti -1c
.RI "std::list< \fBsocket_data_t\fP > \fB_sockdata\fP"
.br
.ti -1c
.RI "std::shared_ptr< \fBkeep_running_flags\fP > \fB_sharedptr_keep_running\fP"
.br
.ti -1c
.RI "std::shared_ptr< \fBconnections\fP > \fB_p_cur_connections\fP"
.br
.ti -1c
.RI "std::shared_ptr< \fBSemaphore\fP > \fB_shpt_semIPCfile\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line \fB50\fP of file \fBthread_pair\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "thread_pair::thread_pair ()\fC [delete]\fP"

.SS "thread_pair::thread_pair (\fBMessageQueue\fP write_queue_id, \fBMessageQueue\fP common_queue, int idx, std::shared_ptr< \fBkeep_running_flags\fP > shpt_keep_running, std::shared_ptr< \fBconnections\fP > shpt_conn, std::shared_ptr< \fBSemaphore\fP > shpt_sem)"

.PP
Definition at line \fB5\fP of file \fBthread_pair\&.cpp\fP\&.
.SS "thread_pair::thread_pair (const \fBthread_pair\fP &)\fC [inline]\fP"

.PP
Definition at line \fB76\fP of file \fBthread_pair\&.h\fP\&.
.SS "thread_pair::thread_pair (\fBthread_pair\fP &&)\fC [delete]\fP"

.SH "Member Function Documentation"
.PP 
.SS "int thread_pair::add_sockdata (\fBsocket_data_t\fP sdt)"

.PP
Definition at line \fB280\fP of file \fBthread_pair\&.cpp\fP\&.
.SS "void thread_pair::Attending_Read_Socket (\fBsocket_data_t\fP & sdt)"

.PP
Definition at line \fB33\fP of file \fBthread_pair\&.cpp\fP\&.
.SS "int thread_pair::get_id () const\fC [inline]\fP"

.PP
Definition at line \fB91\fP of file \fBthread_pair\&.h\fP\&.
.SS "int thread_pair::get_idx () const\fC [inline]\fP"

.PP
Definition at line \fB92\fP of file \fBthread_pair\&.h\fP\&.
.SS "int thread_pair::get_read_pipe () const\fC [inline]\fP"

.PP
Definition at line \fB89\fP of file \fBthread_pair\&.h\fP\&.
.SS "int thread_pair::get_size_of_sock_list () const\fC [inline]\fP"

.PP
Definition at line \fB88\fP of file \fBthread_pair\&.h\fP\&.
.SS "int thread_pair::get_sockdata_list (std::list< \fBsocket_data_t\fP > & lsdt)"

.PP
Definition at line \fB308\fP of file \fBthread_pair\&.cpp\fP\&.
.SS "int thread_pair::get_write_pipe () const\fC [inline]\fP"

.PP
Definition at line \fB90\fP of file \fBthread_pair\&.h\fP\&.
.SS "\fBMessageQueue\fP & thread_pair::get_write_queue ()\fC [inline]\fP"

.PP
Definition at line \fB93\fP of file \fBthread_pair\&.h\fP\&.
.SS "int thread_pair::Getting_Json_Msg_Received (std::string & msgin, \fBprotomsg::st_protomsg\fP & v_protomsg, std::string & msgout)"

.PP
Definition at line \fB93\fP of file \fBthread_pair\&.cpp\fP\&.
.SS "\fBthread_pair\fP & thread_pair::operator= (const \fBthread_pair\fP &)\fC [default]\fP"

.SS "\fBthread_pair\fP & thread_pair::operator= (\fBthread_pair\fP &&)\fC [delete]\fP"

.SS "void thread_pair::Prepare_Msg_Json_To_Send (\fBprotomsg::st_protomsg\fP & v_protomsg, std::string msg, Json::Value & json_msg)"

.PP
Definition at line \fB211\fP of file \fBthread_pair\&.cpp\fP\&.
.SS "void thread_pair::reader_thread (int idx_thp)"

.PP
Definition at line \fB122\fP of file \fBthread_pair\&.cpp\fP\&.
.SS "int thread_pair::remove_sockdata (const \fBsocket_data_t\fP & sdt)"

.PP
Definition at line \fB290\fP of file \fBthread_pair\&.cpp\fP\&.
.SS "void thread_pair::writer_thread (int idx_thp)"

.PP
Definition at line \fB222\fP of file \fBthread_pair\&.cpp\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "std::mutex thread_pair::_accept_mutex\fC [private]\fP"

.PP
Definition at line \fB61\fP of file \fBthread_pair\&.h\fP\&.
.SS "\fBMessageQueue\fP thread_pair::_common_queue\fC [private]\fP"

.PP
Definition at line \fB60\fP of file \fBthread_pair\&.h\fP\&.
.SS "int thread_pair::_idx_thp {\-1}\fC [private]\fP"

.PP
Definition at line \fB57\fP of file \fBthread_pair\&.h\fP\&.
.SS "std::shared_ptr<\fBconnections\fP> thread_pair::_p_cur_connections\fC [private]\fP"

.PP
Definition at line \fB65\fP of file \fBthread_pair\&.h\fP\&.
.SS "int thread_pair::_pipe[2] {\-1,\-1}\fC [private]\fP"

.PP
Definition at line \fB58\fP of file \fBthread_pair\&.h\fP\&.
.SS "std::shared_ptr<\fBkeep_running_flags\fP> thread_pair::_sharedptr_keep_running\fC [private]\fP"

.PP
Definition at line \fB64\fP of file \fBthread_pair\&.h\fP\&.
.SS "std::shared_ptr<\fBSemaphore\fP> thread_pair::_shpt_semIPCfile\fC [private]\fP"

.PP
Definition at line \fB67\fP of file \fBthread_pair\&.h\fP\&.
.SS "std::list<\fBsocket_data_t\fP> thread_pair::_sockdata\fC [private]\fP"

.PP
Definition at line \fB63\fP of file \fBthread_pair\&.h\fP\&.
.SS "\fBMessageQueue\fP thread_pair::_write_queue\fC [private]\fP"

.PP
Definition at line \fB59\fP of file \fBthread_pair\&.h\fP\&.
.SS "std::thread thread_pair::th_r"

.PP
Definition at line \fB53\fP of file \fBthread_pair\&.h\fP\&.
.SS "std::thread thread_pair::th_w"

.PP
Definition at line \fB54\fP of file \fBthread_pair\&.h\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Dispatcher TCP/IP from the source code\&.


src/thread_pair.cpp
#include "thread_pair.h"

using namespace std;

thread_pair::thread_pair(MessageQueue write_queue_id, MessageQueue common_queue, int idx_thp, 
        std::shared_ptr<keep_running_flags> shpt_keep_running, std::shared_ptr<connections> shpt_conn, 
        //std::shared_ptr<signal_synch> shpt_sigsyn, 
        std::shared_ptr<Semaphore> shpt_sem)
{
    // _shpt_sigsyn = shpt_sigsyn;
    _write_queue = write_queue_id;

    assert(_write_queue && "write_queue not operative.");
    assert(common_queue && "read_queue (common queue) not operative.");

    // _accept_mutex = std::make_shared<mutex>();

    _common_queue = common_queue;
    _idx_thp = idx_thp;
    _sharedptr_keep_running = shpt_keep_running;
    _p_cur_connections = shpt_conn;
    _shpt_semIPCfile = shpt_sem;

    if (pipe(_pipe) < 0 ) {
        assert(false && "Cannot create a pipe");
	}

    th_r = std::thread(&thread_pair::reader_thread, this, idx_thp);
    th_w = std::thread(&thread_pair::writer_thread, this, idx_thp);
}


void thread_pair::Attending_Read_Socket(socket_data_t &sdt)
{
    int status;
    Socket sock;
    sock.sock = sdt.sd;
    std::string msgin;
    std::string msgout;

    protomsg::st_protomsg v_protomsg;

    LOG_DEBUG << "Reading from sd " << sdt.sd;

    if((status=sock.socket_read(msgin,1024)) <= 0) 
    {
        LOG_ERROR << "Socket_read error: " << errno << ":" << strerror(errno);
        LOG_ERROR << "Socket sd: " << sdt.sd << " IP:Port: " << 
               inet_ntoa(sdt.sockaddr.sin_addr) << ":" << ntohs(sdt.sockaddr.sin_port);

        remove_sockdata(sdt);
        _p_cur_connections->unregister_conn(sdt.idx_con, *_shpt_semIPCfile);
        close(sdt.sd);
        return;
    }
    sdt.rcvinfo = msgin;

    // TO DO - Check IP is in table. Not needed since AccceptThread checks for valid IP.

    if(!Getting_Json_Msg_Received(msgin, v_protomsg, msgout))
    {
        LOG_ERROR << "Error Message Format (json): " << msgin;
        LOG_ERROR << "Socket sd: " << sdt.sd << " IP:Port: " << 
            inet_ntoa(sdt.sockaddr.sin_addr) << ":" << ntohs(sdt.sockaddr.sin_port);

        // Send back MSGIN original.
        v_protomsg.q_write = _common_queue.getid();  // Not usefull in this case...
        if(!_write_queue.send(&v_protomsg, msgin)) 
        {
            LOG_ERROR << "Impossible to send to Writer_Thread through _write_queue: " << sdt.sd;
            remove_sockdata(sdt);
            _p_cur_connections->unregister_conn(sdt.idx_con, *_shpt_semIPCfile);
            close(sdt.sd);
            return;
        }
    }
    else 
    {
        LOG_DEBUG << "Sending to common queue for TuxCli...";
        // Send MSGOUT to common queue to process by TuxCli 
        v_protomsg.q_write = _write_queue.getid();  // Queue to respond to proper Writer_Thread...
        if(!_common_queue.send(&v_protomsg, msgout)) 
        {
            LOG_ERROR << "Impossible to send to TuxCli through _common_queue: " << sdt.sd;
            remove_sockdata(sdt);
            _p_cur_connections->unregister_conn(sdt.idx_con, *_shpt_semIPCfile);
            close(sdt.sd);
            return;
        }
    }
}

int thread_pair::Getting_Json_Msg_Received(string &msgin, protomsg::st_protomsg &v_protomsg, string &msgout)
{
    Json::Value json_msg;
    stringstream ss(msgin);
    ss >> json_msg;

    v_protomsg.terf = json_msg["TERF"].asInt();
    v_protomsg.terl = json_msg["TERL"].asInt();
    strncpy(v_protomsg.guid, json_msg["GUID"].asCString(), sizeof(v_protomsg.guid)-1);
    strncpy(v_protomsg.pid, json_msg["PID"].asCString(), sizeof(v_protomsg.pid)-1);
    strncpy(v_protomsg.aid, json_msg["AID"].asCString(), sizeof(v_protomsg.aid)-1);
    strncpy(v_protomsg.cabx, json_msg["CABX"].asCString(), sizeof(v_protomsg.cabx)-1);
    msgout = std::string(json_msg["MSG"].asCString());

    v_protomsg.mtype = protomsg::TYPE_NORMAL_MSG;

    LOG_DEBUG << "v_protomsg:" << v_protomsg;

    if (v_protomsg.terf==0 || v_protomsg.terl == 0 || 
        strlen(v_protomsg.guid) == 0 || strlen(v_protomsg.pid) == 0 || 
        strlen(v_protomsg.aid) == 0 || strlen(v_protomsg.cabx) == 0 || 
        msgout.size() == 0)
    {
        return 0;
    }
    
    return 1;
}

void thread_pair::reader_thread(int idx_thp)
{
    list<socket_data_t> lsdt;
    int readpipe=get_read_pipe();
    LOG_DEBUG << "reader_thread " << idx_thp;

    while(_sharedptr_keep_running->_keep_accepting.load()) 
    {
        fd_set readset;
        int nfds, ret_sel;

        /*
        std::unique_lock<std::mutex> lck(_shpt_sigsyn->_cv_mutex);

        auto shpt_keep_running = _sharedptr_keep_running;
        // wait for up to 10 seconds
        _shpt_sigsyn->_cv.wait_for(lck, std::chrono::seconds(10),
              [&shpt_keep_running]() { return !shpt_keep_running->_keep_accepting.load(); });
        */

        //sleep(1);
        // STEP 1 - Getting all sockets descriptors to SELECT.

        get_sockdata_list(lsdt);

        FD_ZERO(&readset);
        nfds = 0; 

        FD_SET(readpipe, &readset);
        nfds = (readpipe<nfds) ? nfds : (readpipe+1);
        
        for(auto it=lsdt.begin(); it != lsdt.end(); it++) 
        {
            FD_SET(it->sd, &readset);
            nfds = (it->sd < nfds) ? nfds : (it->sd+1);
        }

        LOG_DEBUG << "Accepting ndfs=" << nfds << " connections. " << sizeof(readset);

        if((ret_sel=select(nfds, &readset, NULL, NULL, NULL)) < 0 ) 
        {
            int someerror=errno;
            LOG_ERROR << "SELECT FAILED: (" << ret_sel << ") " << strerror(someerror);
            if( someerror == EINTR ) 
                continue;
            return;
        }

        if( FD_ISSET(readpipe, &readset) ) 
        {
            read(readpipe,protopipe::GETPIPEMSG,protopipe::LEN_PIPEMSG);
            if(protopipe::GETPIPEMSG[0]==protopipe::ENDING_PIPE[0]) {
                LOG_DEBUG << "ENDING_PIPE received.";
                continue;   // TO REMOVE. Maybe it needs to finish all messages. 
            }
            else 
                LOG_DEBUG << "WEAKUP_PIPE received.";

        }

        for(auto it=lsdt.begin(); it != lsdt.end(); it++) 
        {
            if(_p_cur_connections->check_obsolete(it->idx_con, *_shpt_semIPCfile) != 0 ) 
            {
                /* Obsolete or some issues */

                LOG_DEBUG << "SD obsolete (deleted): " << it->sd << " IP:Port: " << 
                    inet_ntoa(it->sockaddr.sin_addr) << ":" << ntohs(it->sockaddr.sin_port);
                
                remove_sockdata(*it);
                close(it->sd);
            } 
	        else 
	        if(FD_ISSET(it->sd, &readset) ) 
            {
                
                // Getting message from socket...
                Attending_Read_Socket(*it);

            }
        }

    }

    // pthread_sigmask(SIG_SETMASK, &_shpt_sigsyn->_sigset_old, nullptr);

    LOG_DEBUG << "Ending reader_thread " << idx_thp;
}

void thread_pair::Prepare_Msg_Json_To_Send(protomsg::st_protomsg &v_protomsg, string msg, Json::Value &json_msg)
{
    json_msg["TERF"] = v_protomsg.terf;
    json_msg["TERL"] = v_protomsg.terl;
    json_msg["GUID"] = std::string(v_protomsg.guid);
    json_msg["PID"] = std::string(v_protomsg.pid);
    json_msg["AID"] = std::string(v_protomsg.aid);
    json_msg["CABX"] = std::string(v_protomsg.cabx);
    json_msg["MSG"] = msg;
}

void thread_pair::writer_thread(int idx_thp)
{
    long type;
    std::string msg, msgout;
    Json::Value json_msg;
    protomsg::st_protomsg v_protomsg;
    connection cur_conn;
    Socket send_sock;

    LOG_DEBUG << "writer_thread " << idx_thp;

    while(_sharedptr_keep_running->_keep_working.load()) 
    {
        /*
        std::unique_lock<std::mutex> lck(_shpt_sigsyn->_cv_mutex);

        auto shpt_keep_running = _sharedptr_keep_running;
        // wait for up to 10 seconds
        _shpt_sigsyn->_cv.wait_for(lck, std::chrono::seconds(10),
              [&shpt_keep_running]() { return !shpt_keep_running->_keep_working.load(); });
        */
        //sleep(1);
        // STEP 1 - ReadFromQueue  (idx_con on message)
        
        _write_queue.rcv(&v_protomsg, msg);

        if(v_protomsg.mtype == protomsg::TYPE_ENDING_MSG) 
        {
            LOG_DEBUG << "Received protomsg::TYPE_ENDING_MSG"; 
            continue;
        }

        Prepare_Msg_Json_To_Send(v_protomsg, msg, json_msg);
        
        _p_cur_connections->ending_operation(v_protomsg.idx, *_shpt_semIPCfile, cur_conn);

        // STEP 2 - Send msg to socket 

        send_sock.sock = cur_conn.sd;

        stringstream ss;
        string field;
        ss << json_msg;
    
        while (getline(ss, field)) 
            msgout += field;
    
        LOG_DEBUG << "Sending msgout: " << msgout; 

        send_sock.socket_write(msgout);
        
        msgout.clear();
    }

    //pthread_sigmask(SIG_SETMASK, &_shpt_sigsyn->_sigset_old, nullptr);
    LOG_DEBUG << "Ending writer_thread " << idx_thp;
}

int thread_pair::add_sockdata(socket_data_t sdt)
{
    std::lock_guard<std::mutex> l_guard{_accept_mutex};

    _sockdata.emplace_back(sdt);
    
    LOG_DEBUG << "add_sockdata OK!! " << sdt.sd;
    return 0;
}

int thread_pair::remove_sockdata(const socket_data_t &sdt)
{
    std::lock_guard<std::mutex> l_guard{_accept_mutex};

    std::list<socket_data_t>::iterator findIter = std::find(_sockdata.begin(), _sockdata.end(), sdt);

    if(findIter == _sockdata.end()) {
        LOG_DEBUG << "remove_sockdata NOT FOUND!! " << sdt.sd;
        return -1;
    }
    else {
        socket_data_t item = *findIter;
        _sockdata.remove(item);
        LOG_DEBUG << "remove_sockdata OK!! " << sdt.sd;
        return 0;
    }
}

int thread_pair::get_sockdata_list(list<socket_data_t> &lsdt)
{
    std::lock_guard<std::mutex> l_guard{_accept_mutex};

    lsdt = _sockdata;
    
    LOG_DEBUG << "get_sockdata_list OK!! lsdt.size()=" << lsdt.size(); 

    return 0;
}
